Multi-project dependencies in Xcode

brunobowden opened the pull request:
Multi-project dependencies in Xcode
- j2objcXcode finds all podspecs in dependency tree and applies them to Podfile
- “podFile” => “podfile” to match CocoaPods documentation
- .gitignore the Pods directory for CocoaPods

PodspecTask
- j2objcPodspec separated from j2objcXcode to create podspec
- j2objcAssemble{Debug|Release} now depends on j2objcAssembleResources
- j2objcResources now depends on j2objcPodspec
- Podfile podspec paths are now relative instead of absolute
- Podspec headers moved from public_header_files => HEADER_SEARCH_PATHS

multiProject1 test
- Rename projects and workspace to match FAQ example
- Project with two targets: IOS-APP and IOS-APPTests

Other Changes:
- FAQ: Add .gitignore entry, revise folder structure
- Condense warnings on unclean git and SNAPSHOT build
- Clean up a few redundant imports
- Copyright notices for Xcode project

Fixes #432 - Xcode podspec dependencies
Fixes #502 - Split j2objcPodspec task from j2objcXcode task

brunobowden commented:
@advayDev1 - this still needs some more work but this is close enough to be worth getting your feedback.

advayDev1 commented on FAQ.md:
```diff
@@ -18,6 +18,7 @@ Paste the results below, replacing existing contents.
 - [How do I develop with Xcode?](#how-do-i-develop-with-xcode)
 - [How can I speed up my build?](#how-can-i-speed-up-my-build)
 - [What libraries are linked by default?](#what-libraries-are-linked-by-default)
+- [What should I add to .gitignore?](what-should-i-add-to-gitignore)
```
needs a hashtag

brunobowden replied:
done

advayDev1 commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy:
```diff
@@ -46,13 +47,9 @@ class J2objcPlugin implements Plugin<Project> {
-            project.logger.error('j2objc-gradle plugin was built with local modification.\n' +
-                                 'If you encounter issues, please use an official release from:\n' +
-                                 '    https://github.com/j2objc-contrib/j2objc-gradle/releases')
+            project.logger.error('WARNING: j2objc-gradle plugin was built with local git modification.')
```
why'd you remove the link to known good version?

brunobowden replied:
I should've raised this point up front. After seeing this message so many times, it's become a little irritating to me. The main point though is that likely only developers modifying this plugin are going to see this:

1. The warning should be made more visible so you don't miss it by accident
2. If you're good enough to be modifying the plugin, then you'll know where the plugin came from and how to get an unmodified version. So I feel that the link is redundant. It was better to focus on "1" in having a more visible warning.

advayDev1 replied:
fair, but you can stillachieve this goal by including the link in the single logger.error line

brunobowden replied:
I've added back in the link.

I still think that the only people receiving this warning are getting it because they're technically sophisticated enough to download, build and configure their gradle to get a custom plugin. That's a sophisticated group... the very same group that already knows how to get an official version of the plugin. If there was a culture of passing around plugin files, then an extended warning makes more sense. The culture in this case though is to use gradle "plugin" DSL. So less sophisticated users won't encounter these issue.

advayDev1 commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy:
```diff
@@ -46,13 +47,9 @@ class J2objcPlugin implements Plugin<Project> {
-            project.logger.warn('j2objc-gradle plugin was built outside of an official release.\n' +
-                                'If you encounter issues, please use an official release from:\n' +
-                                '    https://github.com/j2objc-contrib/j2objc-gradle/releases')
+            project.logger.warn('WARNING: j2objc-gradle plugin was built with SNAPSHOT version.')
```
ditto

advayDev1 commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/PodspecTask.groovy:
```diff
@@ -0,0 +1,161 @@
+
+    @VisibleForTesting
+    static void validatePodspecPath(String path, boolean relativeRequired) {
+        if (path.contains('//')) {
```
I'm not sure I understand what the double slash is about

brunobowden replied:
It's a mistake in the path provided to the podspec. For example `/home/dir//dir2`. That's the error it's trying to catch.

brunobowden replied:
Besides, this code was just moved from XcodeTask and hasn't been modified for the most part.

advayDev1 commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy:
```diff
@@ -94,170 +62,157 @@ class XcodeTask extends DefaultTask {
+        logger.debug("Pod exists at path: ${getXcodeProjectDir()}")
+
+
+        // Find for all subprojects as well
```
subprojects means literally projects that have directories under this one.
i presume the logic is for all projects that this project depends on, correct?

if so, you should actually enumerate the j2objcLinkage dependency configuration and find all ProjectDependencies there.

brunobowden replied:
done

advayDev1 commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy:
```diff
@@ -94,170 +62,157 @@ class XcodeTask extends DefaultTask {
+            logger.warn("release: ${podspecTask.getPodspecRelease()}")
+        }
+
+        Set<Project> subprojects = project.getSubprojects()
```
see comments above

advayDev1 commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy:
```diff
@@ -94,170 +62,157 @@ class XcodeTask extends DefaultTask {
+        podspecs.add(j2objcPodspec.getPodspecRelease())
+
+        // Find podspecs from dependencies
+        Set<Object> dependencies = proj.tasks.getByName('j2objcPreBuild').getDependsOn()
```
use the j2objcLinkage dependency configuration instead

advayDev1 commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy:
```diff
@@ -284,13 +239,28 @@ class XcodeTask extends DefaultTask {
     static void writeUpdatedPodfileIfNeeded(
-            File podFile, List<String> xcodeTargets, String podPath,
-            String podNameMethod, String podNameDebug, String podNameRelease, Logger logger) {
+            File podFile, List<File> podspecPairs,
```
please pass a List of pairs or a List of POGOs that specify debug and release files instead

brunobowden replied:
Changed to list of POGOs. I agree with about making the change. It's much better this way.

brunobowden replied:
done

advayDev1 commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy:
```diff
@@ -284,13 +239,28 @@ class XcodeTask extends DefaultTask {
+
+            // 'j2objc_PROJECT' name is generated from the debug/release podspecs
+            // This is hacky but it does work
+            String podNameMethod = Utils.greatestCommonPrefix(podspecDebug.getName(), podspecRelease.getName())
```
just pass in a POGO with debug file, release file, and project name
much easier ot understand

brunobowden replied:
done

advayDev1 commented on systemTests/multiProject1/ios/Podfile:
```diff
@@ -0,0 +1,28 @@
+# platform :ios, '6.0'
+
+def j2objc_extended
+    pod 'j2objc-extended-debug.podspec', :configuration => ['Debug'], :path => '/Users/brunobowden/dev/j2objc-gradle/systemTests/multiProject1/extended/build'
```
how do we avoid checking in stuff with /brunobowden/

brunobowden replied:
I've switch the Podfiles from absolute to relative paths, which eliminate this particular issue. I expect that there will remain issues with the Xcode project itself but I doubt that there's anything I can do about that.

advayDev1 commented on systemTests/multiProject1/ios/Podfile.lock:
```diff
@@ -0,0 +1,39 @@
+  - j2objc-testLib-release (1.0)
+
+DEPENDENCIES:
+  - j2objc-base-debug (from `/Users/brunobowden/dev/j2objc-gradle/systemTests/multiProject1/base/build`)
```
ditto.

advayDev1 commented on systemTests/multiProject1/testLib/build.gradle:
```diff
@@ -28,6 +28,8 @@ dependencies {
 }
 
 j2objcConfig {
+    translateArgs '--no-package-directories'
```
why is this required? also if this CL is forcing this, i think we need to fix that before submitting.

if it isn't required, let's revert to using package-directories.

brunobowden replied:
This is the error that I'm getting. Not sure why...

```
duplicate symbol _OBJC_CLASS_$_ComExampleCubeTesterTest in:
    /Users/brunobowden/dev/j2objc-gradle/systemTests/multiProject1/testLib/build/objs/testJ2objcExecutable/debug/testJ2objcObjc/4wtjh9biqnmeqyjm9sphbjt1c/CubeTesterTest.o
    /Users/brunobowden/dev/j2objc-gradle/systemTests/multiProject1/testLib/build/objs/testJ2objcExecutable/debug/testJ2objcObjc/9zk2mdgcefmbty9x97pvmikua/CubeTesterTest.o
duplicate symbol _OBJC_METACLASS_$_ComExampleCubeTesterTest in:
    /Users/brunobowden/dev/j2objc-gradle/systemTests/multiProject1/testLib/build/objs/testJ2objcExecutable/debug/testJ2objcObjc/4wtjh9biqnmeqyjm9sphbjt1c/CubeTesterTest.o
    /Users/brunobowden/dev/j2objc-gradle/systemTests/multiProject1/testLib/build/objs/testJ2objcExecutable/debug/testJ2objcObjc/9zk2mdgcefmbty9x97pvmikua/CubeTesterTest.o
duplicate symbol _new_ComExampleCubeTesterTest_init in:
    /Users/brunobowden/dev/j2objc-gradle/systemTests/multiProject1/testLib/build/objs/testJ2objcExecutable/debug/testJ2objcObjc/4wtjh9biqnmeqyjm9sphbjt1c/CubeTesterTest.o
    /Users/brunobowden/dev/j2objc-gradle/systemTests/multiProject1/testLib/build/objs/testJ2objcExecutable/debug/testJ2objcObjc/9zk2mdgcefmbty9x97pvmikua/CubeTesterTest.o
duplicate symbol _ComExampleCubeTesterTest_init in:
    /Users/brunobowden/dev/j2objc-gradle/systemTests/multiProject1/testLib/build/objs/testJ2objcExecutable/debug/testJ2objcObjc/4wtjh9biqnmeqyjm9sphbjt1c/CubeTesterTest.o
    /Users/brunobowden/dev/j2objc-gradle/systemTests/multiProject1/testLib/build/objs/testJ2objcExecutable/debug/testJ2objcObjc/9zk2mdgcefmbty9x97pvmikua/CubeTesterTest.o
duplicate symbol _ComExampleCubeTesterTest_class_ in:
    /Users/brunobowden/dev/j2objc-gradle/systemTests/multiProject1/testLib/build/objs/testJ2objcExecutable/debug/testJ2objcObjc/4wtjh9biqnmeqyjm9sphbjt1c/CubeTesterTest.o
    /Users/brunobowden/dev/j2objc-gradle/systemTests/multiProject1/testLib/build/objs/testJ2objcExecutable/debug/testJ2objcObjc/9zk2mdgcefmbty9x97pvmikua/CubeTesterTest.o
ld: 5 duplicate symbols for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

advayDev1 commented:
cool.  2 main comments, rest are minor:
- Use the j2objcLinkage dependencies and/or NativeCompilation.dependsOnProjects instead of decoding the task dependencies.  The dependencies/dependsOnProjects are the spec; while the Task's dependsOn trees are an implementation detail.  They can easily change (I've had to change them in the past already).  Also, (as in my app), users can customize the dependency tree to force things to build in a particular order.
- If this commit is requiring --no-package-directories, we need to fix that.

I assume you're going to take a look at your logs as well before this is ready.

brunobowden commented:
Thanks. I'll take a look at this tomorrow.

What did you mean by "look at my logs"?

On Fri, Oct 9, 2015 at 11:14 PM Advay Mengle <notifications@github.com>
wrote:

> cool. 2 main comments, rest are minor:
>
>    - Use the j2objcLinkage dependencies and/or
>    NativeCompilation.dependsOnProjects instead of decoding the task
>    dependencies. The dependencies/dependsOnProjects are the spec; while the
>    Task's dependsOn trees are an implementation detail. They can easily change
>    (I've had to change them in the past already). Also, (as in my app), users
>    can customize the dependency tree to force things to build in a particular
>    order.
>    - If this commit is requiring --no-package-directories, we need to fix
>    that.
>
> I assume you're going to take a look at your logs as well before this is
> ready.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/j2objc-contrib/j2objc-gradle/pull/504#issuecomment-147045111>
> .
>

advayDev1 commented:
logs: you added a number of logger.error and .warns that should probably be deleted.

brunobowden commented:
@advayDev1 - PTAL. I've addressed the first major issue of using the j2objcLinkage / NativeCompilation. There's a lot of refactoring there which has put it in to a much better place. Still need to test out if no-package-dependencies is really needed. Also, please mark any logger statements that I should remove. Just look at the second commit to make it more manageable for you.

brunobowden commented:
P.S. The podspec paths are now relative rather than absolute, so no more 'brunobowden' in that file. I believe that the other 'brunobowden' mentions in the Xcode project may be unavoidable.

brunobowden commented:
The 'brunobowden' mentions are completely removed.

I'm still having an issue with --no-package-dependencies. I can get the j2objc builds to work. When doing the import in Xcode, it appears that CocoaPods flattens all the header includes. So doing an import of "com/example/Cube.h" fails as it has to look for "Cube.h" instead.

My suggestion is to merge the PR as it stands, then work separately on this issue. Are you ok with that? If so, then do a final pass through the changes and give me an LGTM.

Here's an example of the Header directory as it's copied over to Pods:
```
Brunos-Air:multiProject1 brunobowden$ ls -l ios/Pods/Headers/Public/j2objc-extended-debug
total 480
lrwxr-xr-x  1 brunobowden  staff  104 Oct 10 23:36 $Gson$Preconditions.h -> ../../../../../extended/build/j2objcOutputs/src/main/objc/com/google/gson/internal/$Gson$Preconditions.h
lrwxr-xr-x  1 brunobowden  staff   96 Oct 10 23:36 $Gson$Types.h -> ../../../../../extended/build/j2objcOutputs/src/main/objc/com/google/gson/internal/$Gson$Types.h
lrwxr-xr-x  1 brunobowden  staff  106 Oct 10 23:36 ArrayTypeAdapter.h -> ../../../../../extended/build/j2objcOutputs/src/main/objc/com/google/gson/internal/bind/ArrayTypeAdapter.h
...
lrwxr-xr-x  1 brunobowden  staff   84 Oct 10 23:36 ExtendedCube.h -> ../../../../../extended/build/j2objcOutputs/src/main/objc/com/example/ExtendedCube.h
```

advayDev1 commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcConfig.groovy:
```diff
@@ -386,6 +386,15 @@ class J2objcConfig {
+    List<Project> getBeforeProjects() {
+        return nativeCompilation.beforeProjects
+    }
+    /**
```
newline

brunobowden replied:
done

advayDev1 commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy:
```diff
@@ -250,8 +249,17 @@ class J2objcPlugin implements Plugin<Project> {
                 srcPackedLibDir = file("${buildDir}/packedBinaries/${project.name}-j2objcStaticLibrary")
             }
+            tasks.create(name: 'j2objcPodspec', type: PodspecTask,
+                    // DO NOT COMMIT - separate to debug and release tasks?
```
Can you resolve these DO NOT COMMITs?
whether or not they are actually needed is important to the design.

brunobowden replied:
Resolved

advayDev1 commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/PodspecTask.groovy:
```diff
@@ -0,0 +1,161 @@
+        String libName = "${project.name}-j2objc"
+
+        // podspec creation
+        // TODO: allow custom list of libraries
```
Also add a Todo for handling the OSX (x86_64) directories.

brunobowden replied:
Added TODO

advayDev1 commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/PodspecTask.groovy:
```diff
@@ -0,0 +1,161 @@
+
+        logger.debug("Writing debug podspec... ${getPodspecDebug()}\n$podspecContentsDebug")
+        getPodspecDebug().write(podspecContentsDebug)
+        logger.debug("Writing release podspec... ${getPodspecDebug()}\n$podspecContentsDebug")
```
debug -> release

brunobowden replied:
done

advayDev1 commented on systemTests/multiProject1/ios/IOS-APP/AppDelegate.swift:
```diff
@@ -1,18 +1,10 @@
+//  AppDelegate.swift
+//  IOS-APP
+//
+//  Created by Bruno Bowden on 10/7/15.
```
revert the notice

brunobowden replied:
Added copyright notices to all the files in multiProject1

advayDev1 commented on systemTests/multiProject1/ios/IOS-APP/IOS-APP-Bridging-Header.h:
```diff
@@ -0,0 +1,14 @@
+//
+//  IOS-APP-Bridging-Header.h
+//  IOS-APP
+//
```
notice

advayDev1 commented on systemTests/multiProject1/ios/IOS-APP/ViewController.swift:
```diff
@@ -0,0 +1,32 @@
+//
+//  ViewController.swift
+//  IOS-APP
+//
```
notice

advayDev1 commented on systemTests/multiProject1/ios/IOS-APPTests/IOS_APPTests.swift:
```diff
@@ -0,0 +1,36 @@
+//  IOS_APPTests.swift
+//  IOS-APPTests
+//
+//  Created by Bruno Bowden on 10/7/15.
```
notice

advayDev1 commented:
@brunobowden - the only important thing i'd like to still understand is what you are doing with the DO NOT COMMITs.

re: --no-package-directories, I guess this PR can sustain that, but I don't think we can release 0.5 until that is resolved.

brunobowden commented:
This should now be ready to go.

- Resolve the --no-package-directories by switching the podspecs from public_header_files to HEADER_SEARCH_PATHS

Main thing I'd like feedback on is the dependency changes. I noticed that j2objcPodspec wasn't being built. Curiously, j2objcBuild also doesn't depend on the j2objcAssembleResources tasks, so that was being missed too. I've changed j2objcAssemble{Debug|Release} to depend on j2objcAssembleResources, which in turn depends on j2objcPodspec. You could argue that the podspec is a 'resource to be assembled'. I'm not completely comfortable with that description but it appears the best with the existing dependencies.

advayDev1 commented:
@brunobowden 

1. did you repush?  for example, i'm still seeing all the incorrect copyright notices.

2. AssembleResourcesTask has as its output directory getDestSrcDirFile('main', 'resources').
PodSpecTask has as its input (I know it not an input file) getDestSrcDirFile('main', 'resources').
AssembleResourcesTask always occurs after PodspecTask.
Does gradle incremental build handle this correctly?  If you ./gradlew build twice in a row, on the second run does PodspecTask run (it shouldn't).

LGTM if 1 and 2 are good.

advayDev1 commented:
oh you might want to add 'Handling of multi-project dependencies in Cocoa Pods' to the Changelog

brunobowden commented:
1. I've pushed the changes with the copyright notices. I've also rebased the PR as well.
2. Gradle handles the dependencies fine. The ordering of events doesn't matter, only the @Input and @Output. Since I changed it from @InputDirectory to @Input, it now works just fine. I didn't notice one failed UP-TO-DATE (issue #505), how this is a pre-existing issue.

Given the complexity involved, I'll wait for all the checks to pass before merging.
